### PR TITLE
Pull dependencies implicitly when running a report.

### PR DESCRIPTION
**Pull request by plietar**
If a user runs a report with a `allow_remote=True` in their search options, specifying a dependency will look for packets in remote locations, and pull files from those locations as necessary. If `pull_metadata` is set as well, the metadata will be synchronized from all (enabled) locations.

By default, only the requested files are copied, and not whole packets. However, if the `require_complete_tree` option is enabled on the root then the entire packet, and all of its transitive dependencies, are pulled as well.

### CODE REVIEW

File: src/outpack/filestore.py [hunk 1]
```diff
@@ -22,7 +23,7 @@ def get(self, hash, dst, *, overwrite=False):
         src = self.filename(hash)
         if not os.path.exists(src):
             msg = f"Hash '{hash}' not found in store"
-            raise Exception(msg)
+            raise FileNotFoundError(ENOENT, msg)
```
Comment (richfitz) [inline src/outpack/filestore.py:26]: Is just for convention, or is something else responding more nicely here?

File: src/outpack/location_pull.py [hunk 2]
```diff
@@ -122,16 +121,37 @@ def _mark_all_known(driver, root, location_name):
...
+        actual_options = SearchOptions.create(options)
+
+    if not actual_options.allow_remote:
+        msg = "'allow_remote' must be True"
+        raise Exception(msg)
+
+    if recursive is None:
+        recursive = root.config.core.require_complete_tree
+
+    if root.config.core.require_complete_tree and not recursive:
```
Comment (richfitz) [inline src/outpack/location_pull.py:146]: I presume this duplicates the logic in the R version (so this comment applies to both). I wonder if we're being overly cautious here in the case where we are trying to pull packets that don't _need_ a recursive pull?  In that sense it's only really an issue if we work out we need more careful check that we're not needing to pull more than one id?  We can punt this into a ticket for later easily though

File: src/outpack/location_pull.py [hunk 3]
```diff
@@ -323,21 +375,13 @@ def _location_build_pull_plan(
...
 
-    return LocationPullPlan(packets=fetch, files=files, info=info)
+    return LocationPullPlan(packets=fetch, files=files_location, info=info)
 
 
 def _location_build_pull_plan_packets(
     packet_ids: List[str], root: OutpackRoot, *, recursive: Optional[bool]
 ) -> PullPlanPackets:
     requested = packet_ids
-    if recursive is None:
```
Comment (richfitz) [inline src/outpack/location_pull.py:333]: ah, I see that code has come from here...

File: src/orderly/current.py [hunk 4]
```diff
@@ -63,15 +67,17 @@ def interactive():
             name=path.name,
             id=None,
             orderly=OrderlyCustomMetadata(),
+            # TODO: expose a way of configuring this
```
Comment (richfitz) [inline src/orderly/current.py:70]: The way this is done from R is pretty gross, tbh, so this might be worth discussing generally at some point

File: src/outpack/copy_files.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+def _plan_copy_files(root, id, files):
+    meta = root.index.metadata(id)
+    files = _validate_files(files)
+    known = {f.path: f for f in meta.files}
+
+    plan = {}
+    for here, there in files.items():
+        # TODO: check absolute paths
+        if here.endswith("/") or there.endswith("/"):
+            msg = "Directories not yet supported for export"
```
Comment (richfitz) [inline src/outpack/copy_files.py:81]: Directories are unfortunately quite nasty to get right - but they also seem to be popular with users. Definitely best done in another PR though...

**Review by richfitz — APPROVED** [review]

File: src/outpack/filestore.py (hunk 1 above)
Comment (plietar) [reply src/outpack/filestore.py:26]: I use that to catch the error in `copy_files` to figure out when something is not in the store.

File: src/outpack/location_pull.py (hunk 2 above)
Comment (plietar) [reply src/outpack/location_pull.py:146]: You mean if the packet we pull either has no dependencies, all the dependencies are already satisfied, or they are also specified as part of `ids`? I think that make sense.

File: src/outpack/copy_files.py (hunk 5 above)
Comment (plietar) [reply src/outpack/copy_files.py:81]: That code was just moved from `helpers.py`
